File: C2/PDOO/teoría/preparacion_examen_teoria/Ruby/simple_geometria.rb
module Simple_geometria

#
# ────────────────────────────────────────────────── INTERFACES Y ENUMERADOS ─────
#

    module Form_type
        CIRCULO   = :cirulo
        CUADRADO  = :cuadrado
        TRIANGULO = :triangulo
    end



    class Angulo
        def initialize l1, l2, angulo
            @lado1  = l1
            @lado2  = l2
            @angulo = angulo
        end

        def self.new_specific_angle angulo
            new(0, 0, angulo)
        end

        def self.new_angle_given_sides l1, l2
            ang = Math.atan2(l2, l1)
            new(l1, l2, ang)
        end

        private_class_method :new
        attr_reader :lado1, :lado2, :angulo

        def to_s
            "Lados: " + @lado1.to_s + " " + @lado2.to_s + ". Ángulo: " + @angulo.to_s
        end
    end



    class Geometria
        def area
            0.0
        end

        def angulos
            Array.new
        end

        def perimetro
            0.0
        end
    end


#
# ──────────────────────────────────────────────────── I ──────────
#   :::::: C L A S E S : :  :   :    :     :        :          :
# ──────────────────────────────────────────────────────────────
#


    class Forma < Geometria
        @@formas_creadas = 0

        def initialize
            super()
            @tipo
        end

        attr_reader :tipo

        def self.formas_creadas
            @@formas_creadas
        end
    end



    class Cuadrado < Forma
        def initialize l
            super()
            @tipo = Form_type::CUADRADO
            @@formas_creadas += 1
            @lado = l
        end

        attr_reader :lado

        def area
            @lado * @lado
        end

        def perimetro
            @lado * 4
        end

        def angulos
            angulos = Array.new
            4.times(angulos.push(Angulo.new_angle_given_sides(@lado, @lado)))
            angulos
        end

        def to_s
            "Datos del cuadrado:" + "\n\tLado: " + @lado.to_s + "\n\tÁrea: " + area().to_s + "\n\tPerímetro: " + perimetro().to_s
        end
    end



    class Triangulo < Forma
        def initialize l1, l2, l3
            super()
            @tipo = Form_type::TRIANGULO
            @@formas_creadas += 1

            @lado1 = l1
            @lado2 = l2
            @lado3 = l3

            @angulos = Array.new

            calcular_angulos()
        end

        attr_reader :lado1, :lado2, :lado3, :angulos

        def calcular_angulos
            angulos.push(Angulo.new_angle_given_sides(@lado1, @lado2))
            angulos.push(Angulo.new_angle_given_sides(@lado2, @lado3))
            angulos.push(Angulo.new_angle_given_sides(@lado1, @lado3))
        end

        def perimetro
            @lado1 + @lado2 + @lado3
        end

        def area
            sp = perimetro()/2
            Math.sqrt(sp * (sp - @lado1) * (sp - @lado2) * (sp - @lado3))
        end

        def to_s
            salida = "Datos del triángulo: \n\tLados: " + @lado1.to_s + " " + @lado2.to_s + " " + @lado3.to_s + "\n\tÁrea: " + area().to_s + "\n\tPerímetro: " + perimetro().to_s + "\n\tÁngulos: ["
            @angulos.each{ |angulo|
                salida += angulo.to_s + ", "
            }
            salida += "]"

            salida
        end
    end

    class Circulo < Forma
        def initialize radio
            super()
            @@formas_creadas += 1
            @tipo = Form_type::CIRCULO
            @radio = radio
        end

        attr_reader :radio

        def area
            Math::PI * @radio * @radio
        end

        def perimetro
            2 * Math::PI * @radio
        end

        def to_s
            "Datos del círculo:" + "\n\tRadio: " + @radio.to_s + "\n\tÁrea: " + area().to_s + "\n\tPerímetro: " + perimetro().to_s
        end
    end

    circulo   = Circulo.new(1)
    cuadrado  = Cuadrado.new(3)
    triangulo = Triangulo.new(3, 4, 5)

    puts circulo.area()
    puts Forma.formas_creadas

    puts triangulo, "\n"
    puts cuadrado, "\n"
    puts circulo
end

File: C2/PDOO/teoría/preparacion_examen_teoria/Java/Simple_geometria/src/simple_geometria/simple_geometria.java
package simple_geometria;

import java.util.ArrayList;

// Dejamos las clases y los tipos con visibilidad de paquete (por defecto) por comodidad

//
// ────────────────────────────────────────────────── INTERFACES Y ENUMERADOS ─────
//

    enum form_type {
        CIRCULO,
        CUADRADO,
        TRIANGULO,
    }


    class angulo {
        public double lado1;
        public double lado2;
        public double angulo;

        angulo () {
            lado1 = 0f;
            lado2 = 0f;
            angulo = 0.0;
        }

        angulo (double l1, double l2) {
            lado1 = l1;
            lado2 = l2;

            angulo = Math.atan(lado2/lado1);
        }

        @Override
        public String toString() {
            return "Lados: " + lado1 + " " + lado2 + ". Ángulo: " + angulo;
        }
    }


    interface geometria {
        default public double area () {
            return 0f;
        }

        default public ArrayList<angulo> angulos () {
            return new ArrayList<angulo>();
        }

        default public double perimetro () {
            return 0;
        }

        //public Float altura ();
    }


//
// ──────────────────────────────────────────────────── I ──────────
//   :::::: C L A S E S : :  :   :    :     :        :          :
// ──────────────────────────────────────────────────────────────
//

    abstract class Forma implements geometria {
        protected form_type tipo;
        public static int formas_creadas;

        form_type tipo () {
            return tipo;
        }
    }



    class Cuadrado extends Forma {
        private double lado = 0f;

        Cuadrado () {
            super();
            tipo = form_type.CUADRADO;
            formas_creadas++;
        }

        Cuadrado (double side) {
            super();
            tipo = form_type.CUADRADO;
            lado = side;
            formas_creadas++;
        }

        @Override
        public double area () {
            return lado * lado;
        }

        @Override
        public ArrayList<angulo> angulos () {
            ArrayList<angulo> angulos = new ArrayList<>();

            for (int i = 0; i < 4; i++) {
                angulos.add(new angulo(lado, lado));
            }

            return angulos;
        }

        @Override
        public double perimetro() {
            return 4 * lado;
        }

        /**
         * @return the lado
         */
        public double getLado() {
            return lado;
        }

        @Override
        public String toString() {
            return "Datos del cuadrado:"
                    + "\n\tLado: " + lado
                    + "\n\tÁrea: " + area()
                    + "\n\tPerímetro: " + perimetro();
        }

        // public Float altura () { return lado; }
    }



    class Triangulo extends Forma {
        private double lado1 = 0f;
        private double lado2 = 0f;
        private double lado3 = 0f;

        private ArrayList<angulo> angulos = new ArrayList<>();

        Triangulo (double side1, double side2, double side3) {
            super();
            tipo = form_type.TRIANGULO;
            formas_creadas++;

            lado1 = side1;
            lado2 = side2;
            lado3 = side3;

            calcular_angulos();
        }

        private void calcular_angulos() {
            angulos.add(new angulo(lado1, lado2));
            angulos.add(new angulo(lado2, lado3));
            angulos.add(new angulo(lado1, lado3));
        }

        @Override
        public double perimetro() {
            return lado1 + lado2 + lado3;
        }

        @Override
        public double area() {
            // Fórmula de Heron

            double sp = perimetro() / 2;
            return Math.sqrt(sp * (sp - lado1) * (sp - lado2) * (sp - lado3));
        }

        @Override
        public ArrayList<angulo> angulos() {
            return angulos;
        }

        /**
         * @return the lado1
         */
        public double getLado1() {
            return lado1;
        }

        /**
         * @return the lado2
         */
        public double getLado2() {
            return lado2;
        }

        /**
         * @return the lado3
         */
        public double getLado3() {
            return lado3;
        }

        @Override
        public String toString() {
            String salida =  "Datos del triángulo: \n\tLados: " + lado1 + " " + lado2 + " " + lado3
                        + "\n\tÁrea: " + area()
                        + "\n\tPerímetro: " + perimetro()
                        + "\n\tÁngulos: [";

            for (angulo elemento: angulos)
                salida += elemento.toString() + ", ";

            salida += "]";

            return salida;
        }
    }



    class Circulo extends Forma {
        private double radio = 0;

        Circulo () {
            super();
            formas_creadas++;
            tipo = form_type.CIRCULO;
        }

        Circulo (double radio) {
            super();
            formas_creadas++;
            this.radio = radio;
            tipo = form_type.CIRCULO;
        }

        @Override
        public double area() {
            return Math.PI * radio * radio;
        }

        @Override
        public double perimetro() {
            return 2 * Math.PI * radio;
        }

        /**
         * @return the radio
         */
        public double getRadio() {
            return radio;
        }

        @Override
        public String toString() {
            return "Datos del círculo:"
                    + "\n\tRadio: " + radio
                    + "\n\tÁrea: " + area()
                    + "\n\tPerímetro: " + perimetro();
        }
    }


    class ArrayFormas<T extends Forma> implements geometria {
        private ArrayList<T> figuras;

        ArrayFormas () {
            figuras = new ArrayList<>();
        }

        @Override
        public double area() {
            double resultado = 0;

            for (T elemento: figuras)
                resultado += elemento.area();

            return resultado;
        }

        @Override
        public ArrayList<angulo> angulos() {
            ArrayList<angulo> total = new ArrayList<>();

            for (T elemento: figuras)
                total.addAll(elemento.angulos());

            return total;
        }

        @Override
        public double perimetro() {
            double resultado = 0;

            for (T elemento: figuras)
                resultado += elemento.perimetro();

            return resultado;
        }
    }

//
// ───────────────────────────────────────────────────────────────────── MAIN ─────
//

    public class simple_geometria {
        public static void main(String[] args) {
            Circulo   circulo   = new Circulo(1);
            Cuadrado  cuadrado  = new Cuadrado(3);
            Triangulo triangulo = new Triangulo(3, 4, 5);


            System.out.println(circulo.area());
            System.out.println(Forma.formas_creadas);

            System.out.println(triangulo);
            System.out.println(cuadrado);
            System.out.println(circulo);

            // TODO: practicar polimorfismo, errores de código, probar clase ArrayFormas
        }
    }